airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py

```diff
@@ -0,0 +1,139 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+from datetime import date
+from dateutil.relativedelta import *
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+
+from .google_ads import GoogleAds
+from .utils import Utils
+
+
+def chunk_date_range(start_date: str, conversion_window: Optional[int]) -> Iterable[Mapping[str, any]]:
+    """
+    Returns a list of the beginning and ending timetsamps of each month between the start date and now.
+    The return value is a list of dicts {'date': str} which can be used directly with the Slack API
+    """
+    intervals = []
+
+    # As in to return some state when state in abnormal
+    if start_date > date.today().isoformat():
+        return [{"date": start_date}]
+
+    # applying conversion windoe
+    start_date = date.fromisoformat(
+        start_date) - relativedelta(days=conversion_window)
+    yesterday = date.today() - relativedelta(days=1)
+
+    # Each stream_slice contains the beginning and ending timestamp for a 24 hour period
+    while start_date < yesterday:
+        start = start_date
+        intervals.append({"date": start.isoformat()})
+        start_date = min(yesterday, start_date + relativedelta(months=1))
+
+    return intervals
+
+
+class GoogleAdsStream(Stream, ABC):
+    def __init__(self, config):
+        self.config = config
+        self.google_ads_client = GoogleAds(**config)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return response.next_page_token
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        for result in response:
+            record = GoogleAds.parse_single_result(
+                self.get_json_schema(), result)
+            yield record
+
+    def read_records(
+        self,
+        sync_mode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        start_date, end_date = Utils.get_date_params(
+            stream_slice, self.cursor_field)
+        query = GoogleAds.convert_schema_into_query(
+            self.get_json_schema(), self.name, start_date, end_date)
+
+        pagination_complete = False
+        next_page_token = None
+        while not pagination_complete:
+            response = self.google_ads_client.send_request(
+                query, next_page_token)
+            yield from self.parse_response(response, stream_state=stream_state, stream_slice=stream_slice)
+
+            next_page_token = self.next_page_token(response)
```

Comment:
The [google docs on paging](https://developers.google.com/google-ads/api/docs/reporting/paging) seem to indicate the python client doesn't support manual paging: 
> Note: Python and Ruby client libraries currently do not support manual paging.

Apologies if you already tested but it's not clear from just looking at the PR: Are you sure this is working correctly?

Reply:
Yeah looks like pagination is taken care internally